README.md
iBistu
======

This is the repository for bistu mobile client.

changelog
commit 26e4af9abb364788e33582055ab9b2ccd5b95d09
Author: allen <codeday@qq.com>
Date:   Tue Jul 17 15:17:30 2012 +0800

    check master

commit 24c523bd453965206a83724d57161e3a67d8671d
Author: allen <codeday@qq.com>
Date:   Thu Jul 12 08:31:55 2012 +0800

    ready to finish news and yellowpage

commit 3ee2b3c86a0ca3cea1232ccf0cfbc84c32d4bfe0
Author: mobiledev <mobiledev@soumatoiMac.local>
Date:   Mon Jul 9 12:07:38 2012 +0800

    network module

commit 53bd34f6cbee8c53a47612975f0315f1f51b6a48
Author: allen <codeday@qq.com>
Date:   Mon Jul 9 08:35:20 2012 +0800

    start news and yellowpage modules.

commit a9865e40d0430b0de62405e7c4f1f8d4892cfe54
Author: allen <codeday@qq.com>
Date:   Sat Jul 7 18:03:19 2012 +0800

    ready to start with news

commit ebd1679898d57d0fe536439474329bd7fcad5c13
Author: allen <codeday@qq.com>
Date:   Sun Jul 1 16:12:09 2012 +0800

    add map module

commit bb08a3b483a1ce57cc9ea9bc72a49a8152c9d194
Author: allen <codeday@qq.com>
Date:   Sun Jul 1 12:16:06 2012 +0800

    add map module

commit b7e38259eeb40bc1a1dfdad4e47f37ba352aa31c
Author: allen <codeday@qq.com>
Date:   Wed Jun 13 17:58:34 2012 +0800

    update building and classroom

commit 51f2473ebd4ee70213758c678330b45f8b693b92
Author: allen <codeday@qq.com>
Date:   Sun Jun 3 21:18:37 2012 +0800

    complete the classroom original...

commit 79ffdaac481dcd9315e105ac4d76e6eea9b66123
Author: allen <codeday@qq.com>
Date:   Sat Jun 2 08:05:18 2012 +0800

    ready to add classroom module

commit 091833c7f00d40161c68d818e0ab84a27ff8299e
Author: allen <codeday@qq.com>
Date:   Wed May 30 07:53:13 2012 +0800

    2012/5/30

commit 8ee87569047f9d5b6365cc1018baf4d6f7268f66
Author: allen <codeday@qq.com>
Date:   Wed May 30 07:52:57 2012 +0800

    2012/5/30

commit 329a2d17fcad62a22aa5510065c626b0e5a3fe1e
Author: allen <codeday@qq.com>
Date:   Sat May 26 19:06:49 2012 +0800

    update the major.html & ready to get data of major

commit 051586e8d8ae111000b19f647dff80bcf8aaba44
Author: allen <codeday@qq.com>
Date:   Sun May 20 22:30:08 2012 +0800

    ready to get all data.

commit e075529dc8ec9321d65efbd6efed95df20e92b5f
Author: allen <codeday@qq.com>
Date:   Sun May 20 11:10:05 2012 +0800

    replace the phonegap 1.6

commit 4da7aa3b8d28699ebd965794f2db6f2433416e7c
Author: allen <codeday@qq.com>
Date:   Sun May 20 11:04:17 2012 +0800

    complete college ui and can get data from local database

commit e9291a61d6787a4dd88e19e47e5ae24191d99f62
Author: allen <codeday@qq.com>
Date:   Sat May 19 09:42:15 2012 +0800

    remove the id of the college table

commit ef07aaa0f5b017d98a4eb5a0a846c221b949b538
Author: allen <codeday@qq.com>
Date:   Thu May 17 21:56:10 2012 +0800

    add query.js

commit c264417d95e7409d33075f4cbc9b5c4ab7769941
Author: allen <codeday@qq.com>
Date:   Thu May 17 21:55:26 2012 +0800

    modify the base.js and resolve some problems

commit b508d0594225248ed3123fb51d7dc0928735916d
Author: allen <codeday@qq.com>
Date:   Thu May 17 13:00:30 2012 +0800

    ready to use localStorage

commit 0e0256ca65ace596bf668a62831373c6340aa2ab
Author: allen <codeday@qq.com>
Date:   Sun Apr 22 10:37:28 2012 +0800

    Ajax...

commit 0af1878b286b96fbdcbeaa71e726338af8a26b7c
Merge: 2836301 e1dacc6
Author: allen <codeday@qq.com>
Date:   Sat Apr 21 23:43:26 2012 +0800

    Merge remote-tracking branch 'origin/master'
    
    Conflicts:
    	.gitignore

commit 283630164d155703b21efaa6cd3e64671b70e003
Author: allen <codeday@qq.com>
Date:   Sat Apr 21 23:10:00 2012 +0800

    init reposity

commit e1dacc6975f808a59b53bcbbc159a58583ae9b7d
Author: xionglun <codeday@qq.com>
Date:   Sat Apr 21 07:55:04 2012 -0700

    initial commit


assets/www/js/base.js
/*
* This file is the base javascript file for all other javascript
* files.You should add this file into your project.
* Author: Allen Heavey
* E-mail: ista.allen@gmail.com
* Date: 2012/04/19
* */

/***********************************************************
 * iBistuGap
 * This file is used to create a database!
 * author: Allen
 * Data: 2012/04/16
 *
 ***********************************************************/

/**
 * Bug record!
 * course table seems has error!
 * I will fix it tomorrow!
 * 2012/06/09
 * I have fixed it now!
 *
 *  */

var iBistuDB, updateCollegeFlag = 1;
var databaseExist = window.localStorage.getItem("databaseExist");
var updateAllTables = window.localStorage.getItem("updateAllTables");
var CANUPDATE = true, BASICAL_URL = "", networkState,NETWORK_READY = false;

var Bistu = {
    rootDir:"",
    createRoot:function(){
        
    },
    createNew:function(){
        
    },
    getNetworkState: function(){
        return NETWORK_READY;
    },
    latestUpdate:function(time){
        
    },
    loginStatus:function(){
        var loginOrNot = window.localStorage.getItem("loginStatus");
        
        if(loginOrNot == null || loginOrNot == undefined){
            return false;
        }
        else if(loginOrNot == "true"){
            return true;
        }
        
        return false;
    },
    loginToken:function(){
    		
    		var token = window.localStorage.getItem("loginToken");
    		if(token != null && token != "undefined"){
    			return token;
    		}
    		else{
    			return null;
    		}
    },
    updateCollegeFlag:1,
    updateActivityFlag:1
};

console.log("updateCollegeFlag is " + Bistu.updateCollegeFlag);

/*
 * create a ajax function to get the data frome server!
 * then parse and return the responseText to json!
 * */
function getFromServer(type, url) {
    //If url is null,then return;
    if(url == "" || url == null) {
        return null;
    }

    console.log("type=" + type + "-->url=" + url);

    var xhr = new XMLHttpRequest();
    var resp = null;
    xhr.onreadystatechange = function() {
        if(xhr.readyState == 4) {
            if((xhr.status >= 200 && xhr.status < 300) || xhr.status == 304) {
                try {
                    //xhr.responseText is a string!
                    //If you want to use JSON.parse,then there will be a problem!
                    resp = eval('(' + xhr.responseText + ')');
                    console.log("responseText's Type = " + type + " and length = " + resp.length);
                    // console.log(xhr.responseText);
                    switch(type) {
                        case "college":
                            (function() {
                                iBistuDB.transaction(function(tx) {
                                    tx.executeSql('DROP TABLE IF EXISTS college');
                                    tx.executeSql('create table if not exists college (id INTEGER PRIMARY KEY,collegeName,collegeCode)');
                                    console.log("Start to insert-->" + type);
                                    for(var i = 0, len = resp.length; i < len; i++) {
                                        tx.executeSql('insert into college (collegeName,collegeCode)' + ' values ("' + resp[i].collegeName + '","' + resp[i].collegeCode + '")');
                                    }
                                }, errorCB, successCB);
                            })();
                            break;
                        case "major":
                            (function() {
                                iBistuDB.transaction(function(tx) {
                                    tx.executeSql('drop table if exists major');
                                    tx.executeSql('create table if not exists major (id INTEGER PRIMARY KEY, majorName,majorCode,collegeId)');
                                    console.log("Start to insert-->" + type);
                                    for(var i = 0, len = resp.length; i < len; i++) {
                                        tx.executeSql("insert into major (id,majorName,majorCode,collegeId) values ('" + resp[i].id + "','" + resp[i].majorName + "','" + resp[i].majorCode + "','" + Number(resp[i].collegeId) + "')");
                                    }
                                }, errorCB, successCB);
                            })();
                            break;
                        case "building":
                            (function() {
                                iBistuDB.transaction(function(tx) {
                                    tx.executeSql('drop table if exists building');
                                    tx.executeSql('create table if not exists building (id, buildingCode,buildingName)');
                                    console.log("Start to insert-->" + type);
                                    for(var i = 0, len = resp.length; i < len; i++) {
                                        tx.executeSql('insert into building (id,buildingCode,buildingName) values ("' + resp[i].id + '","' + resp[i].buildingCode + '","' + resp[i].buildingName + '")');
                                    }
                                }, errorCB, successCB);
                            })();
                            break;
                        case "courseDetail":
                            (function() {
                                iBistuDB.transaction(function(tx) {
                                    tx.executeSql("drop table if exists coursedetail");
                                    tx.executeSql("create table if not exists coursedetail (id INTEGER PRIMARY KEY,courseListId,courseTeacher,coursePlace,courseTime)");
                                    console.log("Start to insert-->" + type);
                                    for(var i = 0, len = resp.length; i < len; i++) {
                                        tx.executeSql('insert into coursedetail (courseListId,courseTeacher,coursePlace,courseTime) values ("' + resp[i].courseListId + '","' + resp[i].courseTeacher + '","' + resp[i].coursePlace + '","' + resp[i].courseTime + '")');
                                    }
                                }, errorCB, successCB);
                            })();
                            break;
                        case "courseList":
                            (function() {
                                iBistuDB.transaction(function(tx) {
                                    tx.executeSql('drop table if exists courseList');
                                    tx.executeSql('create table if not exists courseList (id INTEGER PRIMARY KEY,courseCode,courseName,majorId)');
                                    console.log("Start to insert-->" + type);
                                    for(var i = 0, len = resp.length; i < len; i++) {
                                        tx.executeSql("insert into courseList (id,courseCode,courseName,majorId) values ('" + resp[i].id + "','" + resp[i].courseCode + "','" + resp[i].courseName + "','" + resp[i].majorId + "')");
                                    }
                                }, errorCB, successCB);
                            })();
                            break;
                        case "course":
                            (function() {
                                iBistuDB.transaction(function(tx) {
                                    tx.executeSql('drop table if exists course');
                                    tx.executeSql('create table if not exists course (courseName,courseEngName,courseCode,courseInfo,courseXs,courseXf,courseXz,courseLb)');
                                    console.log("Start to insert-->" + type + "  length = " + resp.length);
                                    /***
                                     * Here has a big problem!
                                     * we can't insert table with 3 thousands
                                     * items.
                                     * This problem still don't solve!@2012/05/30
                                     * */
                                    /*
                                     * Bug fixed!!!
                                     * I remove the column courseEngName.then it works.
                                     ***************************************************************/
                                    var courseInfos = "";
                                    for(var i = 0, len = resp.length; i < len; i++) {
                                        courseInfos = resp[i].courseInfo.replace(/\n[\s| ]*\r/g, "");
                                        // courseInfos = '';
                                        tx.executeSql("insert into course (courseName,courseEngName,courseCode,courseInfo,courseXs,courseXf,courseXz,courseLb) values ('" + resp[i].courseName + "','" + "','" + resp[i].courseCode + "','" + courseInfos + "','" + resp[i].courseXs + "','" + resp[i].courseXf + "','" + resp[i].courseXz + "','" + resp[i].courseLb + "')");
                                    }

                                }, function(err) {
                                    console.log("Code is "+err.code);
                                    console.log("Msg is " + err.message);
                                }, function(){
                                    console.log("insert into course success!!!")
                                });
                            })();
                            break;
                        case "classtime":
                            (function() {

                                iBistuDB.transaction(function(tx) {

                                    tx.executeSql('drop table if exists classtime');
                                    tx.executeSql('create table if not exists classtime (id INTEGER PRIMARY KEY, classroomId,date,courseId1,courseId2,courseId3,courseId4,courseId5,courseId6,courseId7,courseId8,courseId9,courseId10,courseId11)');
                                    for(var i = 0, len = resp.length; i < len; i++) {
                                        tx.executeSql("insert into classtime (classroomId,date,courseId1,courseId2,courseId3,courseId4,courseId5,courseId6,courseId7,courseId8,courseId9,courseId10,courseId11) values ('" + resp[i].classroomId + "','" + resp[i].date + "','" + resp[i].courseId1 + "','" + resp[i].courseId2 + "','" + resp[i].courseId3 + "','" + resp[i].courseId4 + "','" + resp[i].courseId5 + "','" + resp[i].courseId6 + "','" + resp[i].courseId7 + "','" + resp[i].courseId8 + "','" + resp[i].courseId9 + "','" + resp[i].courseId10 + "','" + resp[i].courseId11 + "')");
                                    }
                                }, errorCB, successCB);
                            })();
                            break;
                        case "classroom":
                            (function() {

                                iBistuDB.transaction(function(tx) {
                                    tx.executeSql('drop table if exists classroom');
                                    tx.executeSql('create table if not exists classroom (id, roomName,roomCode,buildingId)')
                                    for(var i = 0, len = resp.length; i < len; i++) {
                                        tx.executeSql('insert into classroom (id,roomName,roomCode,buildingId) values ("' + resp[i].id + '","' + resp[i].roomName + '","' + resp[i].roomCode + '","' + resp[i].buildingId + '")');
                                    }
                                }, errorCB, successCB);
                            })();
                            break;
                        default:
                            (function() {
                                console.log("Update error!");
                            })();
                    }

                }
                catch(e) {
                    console.log("insert into Table error-->" + type + "-- Type-->" + e.fileName);
                }
            }
            else {
                console.log("Get data from server error code = " + xhr.status);
            }
        }
    }
    xhr.open("GET", url, true);
    xhr.send(null);
}

document.addEventListener("deviceready", onDeviceReadyNow, false);

function checkConnection() {
    networkState = navigator.network.connection.type;
    console.log("networkstate is " + networkState);
    console.log("network's on " + navigator.platform);
    
    if(networkState != "none"){
    		NETWORK_READY = true;
    		
    		if(updateCollegeFlag == 1){
    		    updateCollegeTable();
                updateBuildingTable();
    		}
    }
    else{
    		NETWORK_READY = false;
    }
    
}

function onDeviceReadyNow() {
    checkConnection();
    window.requestFileSystem(LocalFileSystem.PERSISTENT, 0, function(fs){
        Bistu.rootDir = fs.root.fullPath;
        fs.root.getDirectory("iBistu",{create: true, exclusive: false},function(){
            console.log("directory iBistu has create");
        },function(){
            console.log("directory iBistu can't create");
        });
    },function(code){
        console.log("get local file system error!");
    });
}

/*
 * create database or open the database;
 * */
(function() {
    /***************************************************************************
     * 创建数据库，
     ***************************************************************************/
    iBistuDB = window.openDatabase("iBistu", "0.1", "BistuDB", 100000);

    if(iBistuDB != undefined) {
        databaseExist = true;
        window.localStorage.setItem("databaseExist", "true");
    }
    else {
        databaseExist = false;
        window.localStorage.setItem("databaseExit", "false");
    }
    
    if(databaseExist == false) {
        iBistuDB.transaction(populateDB, errorCB, successCB);
    }
})();

var insertFlag = -1;

function createTableOrNot(tx, results) {
    if(results.rowsAffected == 0) {
        insertFlag = 0;
    }
    else {
        insertFlag = 1;
    }
    console.log("createTable flag--->" + insertFlag);
}

// 数据库操作(事务处理)
// 初始化各个表(如果表格不存在的话)
function populateDB(db) {

    db.executeSql('CREATE TABLE IF NOT EXISTS building (id INTEGER PRIMARY KEY, buildingCode,buildingName)', [], createTableOrNot);
    db.executeSql('CREATE TABLE IF NOT EXISTS classroom (id INTEGER PRIMARY KEY, roomName,roomCode,buildingId)', [], createTableOrNot);
    db.executeSql('CREATE TABLE IF NOT EXISTS college (id INTEGER PRIMARY KEY,collegeName,collegeCode)', [], createTableOrNot);
    db.executeSql('CREATE TABLE IF NOT EXISTS course (id INTEGER PRIMARY KEY, courseName,courseEngName,courseCode,courseInfo,courseXs,courseXf,courseXz,courseLb)', [], createTableOrNot);
    db.executeSql('CREATE TABLE IF NOT EXISTS coursedetail (id INTEGER PRIMARY KEY,courseListId,courseTeacher,coursePlace,courseTime)', [], createTableOrNot);
    db.executeSql('CREATE TABLE IF NOT EXISTS courseList (id INTEGER PRIMARY KEY,courseCode,courseName,majorId)', [], createTableOrNot);
    db.executeSql('CREATE TABLE IF NOT EXISTS major (id INTEGER PRIMARY KEY, majorName,majorCode,collegeId)', [], createTableOrNot);
    db.executeSql('CREATE TABLE IF NOT EXISTS favorCourses (id INTEGER PRIMARY KEY, firstPart,secondPart)', [], createTableOrNot);
    db.executeSql('CREATE TABLE IF NOT EXISTS classtime (id INTEGER PRIMARY KEY, classroomId,date,courseId1,courseId2,courseId3,courseId4,courseId5,courseId6,courseId7,courseId8,courseId9,courseId10,courseId11)', [], createTableOrNot);

    console.log("create tables because it's not exist!");
}

// 当执行SQL失败后调用此方法
function errorCB(error) {
    console.log("executeSql error");
}

// 当执行SQL成功后调用此方法
function successCB() {
    console.log("executeSql success");
}

// var counter = 1;
//
// //'INSERT INTO iBistu (id, data) VALUES (2, "Second row")'
// function executeAddSql(db) {
// //这里有一处调试了很久的BUG------> "+ "'"+ inputString + "'" +"
// <-----------必须形如这样的才能被添加进数据库
// //即须加单引号.经网上查询得，如果不加引号或空格，将会把逗号与字符串等看成一个字符串。
// db.executeSql("INSERT INTO iBistu (id, data) VALUES (" + counter + ",'" +
// inputString + "')");
// counter++;
// }

//**************************************************************

function updateBuildingTable() {

    var url = "http://m.bistu.edu.cn/api/api.php?table=building";
    var type = "building";
    getFromServer(type, url);

}

function updateCourseDetailTable() {
    var url = "http://m.bistu.edu.cn/api/api.php?table=coursedetail";
    var type = "courseDetail";
    getFromServer(type, url);
}

function updateCourseListTable() {
    var url = "http://m.bistu.edu.cn/api/api.php?table=courselist";
    var type = "courseList";
    getFromServer(type, url);

}

function updateClassroomTable() {

    var url = "http://m.bistu.edu.cn/api/api.php?table=classroom";
    var type = "classroom";
    getFromServer(type, url);

}

function updateCollegeTable() {
    var url = "http://m.bistu.edu.cn/api/api.php?table=college";
    var type = "college";
    getFromServer(type, url);

}

function updateCourseTable() {
    var url = "http://m.bistu.edu.cn/api/api.php?table=course";
    var type = "course";
    getFromServer(type, url);

}

function updateMajorTable() {
    var url = "http://m.bistu.edu.cn/api/api.php?table=major";
    var type = "major";
    getFromServer(type, url);
}

function updateClasstimeTable() {
    var url = "http://m.bistu.edu.cn/api/api.php?table=classtime";
    var type = "classtime";

    getFromServer(type, url);

}
